server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -155,6 +169,17 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         }
     }
 
+    @Override
+    public RoutingExplanations execute(RoutingAllocation allocation, AllocationCommands commands, boolean explain, boolean retryFailed) {
+        var explanations = ShardsAllocator.super.execute(allocation, commands, explain, retryFailed);
+        for (AllocationCommand command : commands.commands()) {
+            if (command instanceof MoveAllocationCommand move) {
+                pendingDesiredBalanceMoves.add(move);
```

Comment:
Is there the possibility of weird stuff happening because we don't add the whole batch of commands to the list at once? We might execute some of the commands, rebalance a bit, and then execute some more commands. Maybe it should be a `Queue<List<MoveAllocationCommand>>` instead?